hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemChooseSAS.java

```diff
@@ -0,0 +1,145 @@
+   * @throws Exception
+   */
+  @Test
+  public void testBothProviderFixedTokenConfigured() throws Exception {
+    AbfsConfiguration testAbfsConfig = getConfiguration();
+
+    // configuring a SASTokenProvider class: this provides a user delegation SAS
+    // user delegation SAS Provider is set
+    // This easily distinguishes between results of filesystem level and blob level operations to ensure correct SAS is chosen,
+    // when both a provider class and fixed token is configured.
+    testAbfsConfig.set(FS_AZURE_SAS_TOKEN_PROVIDER_TYPE, "org.apache.hadoop.fs.azurebfs.extensions.MockDelegationSASTokenProvider");
+
+    // configuring the fixed SAS token
+    testAbfsConfig.set(FS_AZURE_SAS_FIXED_TOKEN, accountSAS);
+
+    // creating a new fs instance with the updated configs
+    AzureBlobFileSystem newTestFs = (AzureBlobFileSystem) FileSystem.newInstance(testAbfsConfig.getRawConfiguration());
+
+    // testing a file system level operation
+    TracingContext tracingContext = getTestTracingContext(newTestFs, true);
+    // expected to fail in the ideal case, as delegation SAS will be chosen, provider class is given preference when both are configured
+    // this expectation is because filesystem level operations are beyond the scope of Delegation SAS Token
+    intercept(SASTokenProviderException.class,
+        () -> {
+          newTestFs.getAbfsStore().getFilesystemProperties(tracingContext);
+        });
+
+    // testing blob level operation to ensure delegation SAS token is otherwise valid and above operation fails only because it is fs level
+    Path testPath = new Path("/testCorrectSASToken");
+    newTestFs.create(testPath).close();
+  }
+
+  /**
+   * Tests the scenario where only the fixed token is configured, and no token provider class is set:
+   * whether fixed token is read correctly from configs, and whether the chosen SAS Token works as expected
+   * @throws IOException
+   */
+  @Test
+  public void testOnlyFixedTokenConfigured() throws IOException {
+    AbfsConfiguration testAbfsConfig = getConfiguration();
+
+    // clearing any previously configured SAS Token Provider class
+    testAbfsConfig.unset(FS_AZURE_SAS_TOKEN_PROVIDER_TYPE);
+
+    // setting an account SAS token in the fixed token field
+    testAbfsConfig.set(FS_AZURE_SAS_FIXED_TOKEN, accountSAS);
+
+    // creating a new FS with updated configs
+    AzureBlobFileSystem newTestFs = (AzureBlobFileSystem) FileSystem.newInstance(testAbfsConfig.getRawConfiguration());
+
+    // attempting an operation using the selected SAS Token
+    // as an account SAS is configured, both filesystem level operations (on root) and blob level operations should succeed
+    try {
+      newTestFs.getFileStatus(new Path("/"));
+      Path testPath = new Path("/testCorrectSASToken");
+      newTestFs.create(testPath).close();
+      newTestFs.delete(new Path("/"), true);
```

Comment:
recursive root delete is a funny one. what does abfs do here? does it delete everything? I'm curious now. (s3a fs returns false before even trying to talk to the store).

Reply:
In case of delete on a root path, ABFS list down all the children of root and delete them individually. In case a child is itself a directory, it will be deleted recursively.

Reply:
ok. I think for s3a we decided against that on the basis that it was pretty dangerous to do accidentally. nobody ever does "rm -rf /" after all. At least not more than once...